File: 146-lru-cache/lru-cache.java
class LRUCache {
    public int capacityG;
    public Map<Integer, Node> map = new HashMap<>();
    public Node head;
    public Node tail;
    class Node{
        int val;
        int key;
        Node next;
        Node prev;
        public Node(){
            this.next = null;
            this.prev = null;
            this.val = 0;
        }
    }

    public LRUCache(int capacity) {
        this.capacityG = capacity;
        head = new Node();
        tail = new Node();
        head.next = tail;
        tail.prev = head;
    }
    
    public int get(int key) {
        if(map.containsKey(key)){
            remove(map.get(key));
            add(map.get(key));
            return map.get(key).val;
        }
        else{
            return -1;
        }
    }
    
    public void put(int key, int value) {
        if(map.containsKey(key)){
            Node currNode = map.get(key);
            currNode.val = value;
            remove(map.get(key));
            add(map.get(key));
        }
        else{
            if(capacityG == 0){
                map.remove(tail.prev.key);
                remove(tail.prev);
                capacityG+=1;
            }
            Node newNode = new Node();
            newNode.val = value;
            newNode.key = key;
            map.put(key, newNode);
            add(newNode);
            capacityG-=1;
        }
    }
    public void add(Node node){
        node.next = head.next;
        head.next.prev = node;
        head.next = node;
        node.prev = head;
    }
    public void remove(Node node){
        Node prevNode = node.prev;
        Node nextNode = node.next;
        prevNode.next = nextNode;
        nextNode.prev = prevNode;

    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 146-lru-cache/lru-cache.py
class LRUCache:
    class DLL:
        class Node:
            def __init__(self, left, right, val, key):
                self.val = val
                self.key = key
                self.right = right
                self.left = left

        def __init__(self):
            self.head = self.Node(None, None, None, None)
            self.tail = self.Node(None, None, None, None)
            self.head.right = self.tail
            self.tail.left = self.head

        def remove(self):
            deletedKey = self.tail.left.key
            prev = self.tail.left.left
            prev.right = self.tail
            self.tail.left = prev
            return deletedKey

        def add(self, value, key):
            node = self.Node(None, None, value, key)
            nextNode = self.head.right
            self.head.right = node
            node.left = self.head
            node.right = nextNode
            nextNode.left = node
            return node

        def removeNode(self, node):
            nextNode = node.right
            prevNode = node.left
            prevNode.right = nextNode
            nextNode.left = prevNode

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.m = {}
        self.DLL = self.DLL()

    def get(self, key: int) -> int:
        if key not in self.m:
            return -1
        node = self.m[key]
        self.DLL.removeNode(node)
        newNode = self.DLL.add(node.val, node.key)
        self.m[key] = newNode

        return newNode.val

    def put(self, key: int, value: int) -> None:
        if key in self.m:
            self.m[key].val = value
            self.DLL.removeNode(self.m[key])
            newNode = self.DLL.add(value, key)
            self.m[key] = newNode
            return

        if len(self.m) == self.capacity:
            deletedKey = self.DLL.remove()
            del self.m[deletedKey]

        newNode = self.DLL.add(value, key)
        self.m[key] = newNode